### ISSUE TITLE
Support built-in functions

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/terraform-ls#37]
## Background

Terraform configuration language has a number of [built-in functions](https://www.terraform.io/docs/configuration/functions.html), some of which come from HCL.

### Name Completion

When a user requests completion on the RHS (after "="), aside from other candidates (such as references), they're presented with function names compatible with the attribute type, i.e. functions which return `string` (or "any type") if the attribute is of type `string`.

<img width="863" alt="image5" src="https://user-images.githubusercontent.com/287584/162210695-0210281a-45ba-48af-b82f-253d93bc0ab6.png">

### Argument Completion

When the user confirms the function name from completion or types the opening brace, they are provided with any references of the argument type, i.e. if the argument being completed is of `string` type, then any references of `string` type are provided.

<img width="613" alt="image4" src="https://user-images.githubusercontent.com/287584/162210858-4e967146-abc5-466f-9af3-61463664d9af.png">

### Hover

When the user hovers over the function name, they are provided with the signature of the function, brief description and link to the documentation for that function.

<img width="527" alt="image2" src="https://user-images.githubusercontent.com/287584/162211023-e40e7733-30d4-4a33-8455-b92b73f1d941.png">

### Semantic Tokens

The name of a valid function, such as `format` will be highlighted semantically as function name, name of an invalid function (such as `frmt`) will not. That way users can tell the difference between typo-ed function name and correct one just from the colours.

## Implementation Notes

 - [x] `zclconf/go-cty`
    - [x] Introduce `Description` to the [`Spec`](https://pkg.go.dev/github.com/zclconf/go-cty@v1.10.0/cty/function#Spec) and [`Parameter`](https://pkg.go.dev/github.com/zclconf/go-cty@v1.10.0/cty/function#Parameter) types of `zclconf/go-cty/function`
    - [x] Backfill descriptions of functions & args in cty [`stdlib`](https://pkg.go.dev/github.com/zclconf/go-cty@v1.10.0/cty/function/stdlib) from [docs](https://www.terraform.io/language/functions)
 - [x] `hashicorp/terraform`
    - [x] Backfill descriptions of functions & args in [Terraform Core](https://github.com/hashicorp/terraform/blob/main/internal/lang/functions.go) from [docs](https://www.terraform.io/language/functions) to enable signatures to be exposed via CLI later
 - [x] `hashicorp/terraform-schema`
    - [x] Introduce `Signature` (see below) to `terraform-schema` to loosely mimic cty's `function.Spec`
    - [x] Codify all TF functions for 0.12+ versions into `terraform-schema/internal/funcs/functions.go` using the `Signature` type; see [prototype](https://gist.github.com/radeksimko/83aa5f5db4a1ee723af726cc6a83973e) for finding return types
    - [x] Expose `FunctionsForVersion(v *version.Version) (map[string]Signature, error)` from `terraform-schema/schema`
 - [x] `hashicorp/hcl-lang`
    - [x] Add `Functions map[string]function.Function` to [PathContext](https://github.com/hashicorp/hcl-lang/blob/191cd51dec5ba732fc518a05ead520ee7e23bbfb/decoder/path_context.go#L12-L21) in hcl-lang
    - [x] Update `PathDecoder.CandidatesAtPos` in `hcl-lang/decoder`
    - [x] Update `PathDecoder.HoverAtPos` in `hcl-lang/decoder`
    - [x] Update `PathDecoder.SemanticTokensInFile` in `hcl-lang/decoder`
 - [x] `hashicorp/terraform-ls`
    - [x] Call new public `schema.FunctionsForVersion()` from [within the internal/decoder package](https://github.com/hashicorp/terraform-ls/blob/cbc8cd2d8e27e751b1615c3428a268244df1e960/internal/decoder/decoder.go#L29-L34)
    - [x] Introduce a new handler for [`textDocument/signatureHelp`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_signatureHelp) which calls the new `PathDecoder.SignatureAtPos()` method

```go
type FuncSignature struct {
   Name            string
   Description     lang.MarkupContent
   Parameters      []FuncParameter
   ActiveParameter uint
}
 
type FuncParameter struct {
   Name        string
   Description lang.MarkupContent
}
```

```go
type Signature struct {
   Params      []function.Parameter
   VarParam    *function.Parameter
   ReturnTypes []cty.Type
   Description string
}
```

**Comment by danishprakash** [comment]
@radeksimko sorry for the noise but is this something I can work on? thanks

**Comment by radeksimko** [comment]
This will involve some upfront work on Terraform Core (to expose functions in some way) and drafting of the design (weighing pros and cons of different approaches), so it's probably not the best issue to pick for someone outside of HashiCorp _at this point_.

It is possible that once the upstream work is finished we could use some extra hands. It is unlikely that there will be progress on this particular issue in the next 3 weeks though.

**Comment by radeksimko** [comment]
Support for functions was implemented as part of the following PRs:

 - https://github.com/hashicorp/hcl-lang/pull/207
 - https://github.com/hashicorp/hcl-lang/pull/232
 - https://github.com/hashicorp/terraform-schema/pull/145
 - https://github.com/hashicorp/terraform-ls/pull/1077
 - https://github.com/hashicorp/hcl-lang/pull/135

This was all released as part of [v0.31.0](https://github.com/hashicorp/terraform-ls/releases/tag/v0.31.0).

There are some minor related issues to follow-up on which are being tracked separately:

 - https://github.com/hashicorp/terraform-ls/issues/1252
 - https://github.com/hashicorp/terraform-ls/issues/1248
 - https://github.com/hashicorp/terraform-ls/issues/1249
 - https://github.com/hashicorp/vscode-terraform/issues/1388
 - https://github.com/hashicorp/hcl-lang/issues/235

All the features shown in the initial post are however implemented, so I'm going to close this.
